src/calcium/qqbar.jl

```diff
@@ -353,109 +353,41 @@ end
-  return z
-end
-
-function -(a::QQBarFieldElem, b::ZZRingElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_sub_fmpz(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Ref{ZZRingElem})::Nothing
-  return z
-end
-
-function -(a::QQBarFieldElem, b::Int)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_sub_si(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Int)::Nothing
-  return z
-end
-
-function -(a::QQFieldElem, b::QQBarFieldElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_fmpq_sub(z::Ref{QQBarFieldElem}, a::Ref{QQFieldElem}, b::Ref{QQBarFieldElem})::Nothing
-  return z
-end
-
-function -(a::ZZRingElem, b::QQBarFieldElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_fmpz_sub(z::Ref{QQBarFieldElem}, a::Ref{ZZRingElem}, b::Ref{QQBarFieldElem})::Nothing
-  return z
-end
-
-function -(a::Int, b::QQBarFieldElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_si_sub(z::Ref{QQBarFieldElem}, a::Int, b::Ref{QQBarFieldElem})::Nothing
-  return z
-end
-
-function *(a::QQBarFieldElem, b::QQBarFieldElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_mul(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Ref{QQBarFieldElem})::Nothing
+function ^(a::QQBarFieldElem, b::QQBarFieldElem)
+  ok, z = pow!(QQBarFieldElem(), a, b)
+  !ok && throw(DomainError((a, b)))
   return z
 end
 
-function *(a::QQBarFieldElem, b::QQFieldElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_mul_fmpq(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Ref{QQFieldElem})::Nothing
-  return z
-end
+###############################################################################
+#
+#   Ad hoc binary operators
+#
+###############################################################################
 
-function *(a::QQBarFieldElem, b::ZZRingElem)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_mul_fmpz(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Ref{ZZRingElem})::Nothing
-  return z
-end
+for T in (QQFieldElem, ZZRingElem, Int, UInt, Integer, Rational)
+  @eval begin
+    +(a::QQBarFieldElem, b::$T) = add!(QQBarFieldElem(), a, b)
+    +(a::$T, b::QQBarFieldElem) = add!(QQBarFieldElem(), a, b)
 
-function *(a::QQBarFieldElem, b::Int)
-  z = QQBarFieldElem()
-  @ccall libflint.qqbar_mul_si(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Int)::Nothing
-  return z
-end
+    -(a::QQBarFieldElem, b::$T) = sub!(QQBarFieldElem(), a, b)
+    -(a::$T, b::QQBarFieldElem) = sub!(QQBarFieldElem(), a, b)
 
-*(a::QQFieldElem, b::QQBarFieldElem) = b * a
-*(a::ZZRingElem, b::QQBarFieldElem) = b * a
-*(a::Int, b::QQBarFieldElem) = b * a
+    *(a::QQBarFieldElem, b::$T) = mul!(QQBarFieldElem(), a, b)
+    *(a::$T, b::QQBarFieldElem) = mul!(QQBarFieldElem(), a, b)
 
-function ^(a::QQBarFieldElem, b::QQBarFieldElem)
-  z = QQBarFieldElem()
-  ok = Bool(@ccall libflint.qqbar_pow(z::Ref{QQBarFieldElem}, a::Ref{QQBarFieldElem}, b::Ref{QQBarFieldElem})::Cint)
-  !ok && throw(DomainError((a, b)))
-  return z
+    function ^(a::QQBarFieldElem, b::$T)
```

Comment:
```suggestion
    if T !== QQBarFieldElem
      function ^(a::QQBarFieldElem, b::$T)
```

Reply:
I don't understand this comment, `T` will never be `QQBarFieldElem`, as `for T in (QQFieldElem, ZZRingElem, Int, UInt, Integer, Rational)` in line 374

Reply:
I misread. I was looking for the error and thought it was here.